Llava 1.6 support

cmp-nct opened the pull request:
First steps - I got impressive results with llava-1.6-13B on the license_demo example already, despite many open issues.

Todo:

The biggest and most important difference missing is the "spatial_unpad" logic.
The conversion script I added can convert the nested array into a flat 2D array with valid image shapes, but it ignores them at this point.
The new tensor for the image separation is part of the projector - for compatibility with pytorch I removed it from the llava.clip extract.


llava-surgery-v2.py should be compatible with cogvlm, llava-1.6 and llava-1.5


**For Mistral and using llava-cli binary:**
Add this: `-p "<image>\nUSER:\nProvide a full description.\nASSISTANT:\n"`
The mistral template for llava-1.6 seems to be no system print and a USER/ASSISTANT role

**For Vicunas the default settings work.**

**For the 34B this should work:**
Add this: `-e -p <|im_start|>system\nAnswer the questions.<|im_end|><|im_start|>user\n<image>\nWhat can be said about this image?<|im_end|><|im_start|>assistant\n`


Do not expect great results before the proper image preprocessing was added


**Downloads:**
I've extracted the embedded vit and quantized it for all 4 variants (though, not all quantizations)
They are being uploaded here: https://huggingface.co/cmp-nct/llava-1.6-gguf/upload/main

Please note: Until preprocessing is done, expect poor results

cmp-nct committed:
Create llava-survery-v2.py

cmp-nct committed:
Update convert-image-encoder-to-gguf.py

cmp-nct committed:
Update convert-image-encoder-to-gguf.py

cmp-nct committed:
Rename llava-survery-v2.py to llava-surgery-v2.py

cjpais commented:
one note is that for `liuhaotian/llava-v1.6-mistral-7b` the "model.mm_projector" is in `model-00003-of-00004` from my testing

cmp-nct committed:
Update convert-image-encoder-to-gguf.py

will now search for projector

cmp-nct commented:
> one note is that for `liuhaotian/llava-v1.6-mistral-7b` the `model.mm_projector` is in `model-00003-of-00004`
> 
> this impacts the current code as it only looks for mmproj in the last of the `.safetensors` files

I've updated it with a quick solution to search for those two checkpoint paths.
As long as projectors are not split through multiple files it should work now

```
Cleaning Q:\models\llava\llava-v1.6-vicuna-13b\model-00005-of-00006.safetensors
Searching for vision tower tensors in Q:\models\llava\llava-v1.6-vicuna-13b\model-00005-of-00006.safetensors
No vision tower found in Q:\models\llava\llava-v1.6-vicuna-13b\model-00005-of-00006.safetensors
Done! All vision tower tensors are removed from the model files and stored in llava.clip file.
Taking projector from Q:\models\llava\llava-v1.6-vicuna-13b\model-00006-of-00006.safetensors
Taking newline from Q:\models\llava\llava-v1.6-vicuna-13b\model-00001-of-00006.safetensors
Found 4 tensors to extract.
Found additional 1 tensors to extract.
Done!
Now you can convert Q:\models\llava\llava-v1.6-vicuna-13b\ to a a regular LLaMA GGUF file.
Also, use Q:\models\llava\llava-v1.6-vicuna-13b\/llava.projector to prepare a llava-encoder.gguf file.
```

cmp-nct committed:
Update convert-image-encoder-to-gguf.py

whoops

cmp-nct committed:
Update llava-surgery-v2.py

chigkim commented:
It looks like Ollama that uses Llama.cpp as their backend already supports llava:34b-v1.6.
https://ollama.ai/library/llava/tags
Not sure if I can tag people from other repo, but @jmorganca, did you guys modify your own Llama.cpp already to support v1.6 architecture?

cmp-nct commented:
With these tools you can convert llava-1.6 into a llama.cpp GGUF file and it will work for inferencing.
But as long as the image preprocessing is not integrated, it will not provide the same quality in results. 

Right now llama.cpp will create the usual 14 patches of a rectangular padded 336 pixel image.
But the big change in llava-1.6 was the preprocessing in how patches are split up into image regions of much higher resolutions, it does not need the padding/cropping anymore.

chigkim commented:
> With these tools you can convert llava-1.6 into a llama.cpp GGUF file and it will work for inferencing. But as long as the image preprocessing is not integrated, it will not provide the same quality in results.
> 
> Right now llama.cpp will create the usual 14 patches of a rectangular padded 336 pixel image. But the big change in llava-1.6 was the preprocessing in how patches are split up into image regions of much higher resolutions, it does not need the padding/cropping anymore.

Yeah, I get that, but I was wondering if Ollama forked Llama.cpp they're using and already completed their own implementation to match Llava 1.6 architecture.

ggerganov commented:
If it is just the pre-processing missing, we can merge this and make a separate issue with a specific goal to implement that pre-processing. Might be a good idea to first confirm that using a correctly pre-processed image (from the reference implementation) yields good results using this code

chigkim commented:
Basically resize logic from clip_image_preprocess function in clip.cpp needs to be modified to match process_image function in conversation.py?
I need to look more carefully, but it looks like clip_image_preprocess makes the image into square with padding first and resizes to vision_model.hparams.image_size with linear interpolation.

```LLaVA/llava/conversation.py
def process_image(self, image, image_process_mode, return_pil=False, image_format='PNG', max_len=1344, min_len=672):
	if image_process_mode == "Pad":
		def expand2square(pil_img, background_color=(122, 116, 104)):
			width, height = pil_img.size
			if width == height:
				return pil_img
			elif width > height:
				result = Image.new(pil_img.mode, (width, width), background_color)
				result.paste(pil_img, (0, (width - height) // 2))
				return result
			else:
				result = Image.new(pil_img.mode, (height, height), background_color)
				result.paste(pil_img, ((height - width) // 2, 0))
				return result
		image = expand2square(image)
	elif image_process_mode in ["Default", "Crop"]:
		pass
	elif image_process_mode == "Resize":
		image = image.resize((336, 336))
	else:
		raise ValueError(f"Invalid image_process_mode: {image_process_mode}")
	if max(image.size) > max_len:
		max_hw, min_hw = max(image.size), min(image.size)
		aspect_ratio = max_hw / min_hw
		shortest_edge = int(min(max_len / aspect_ratio, min_len, min_hw))
		longest_edge = int(shortest_edge * aspect_ratio)
		W, H = image.size
		if H > W:
			H, W = longest_edge, shortest_edge
		else:
			H, W = shortest_edge, longest_edge
		image = image.resize((W, H))
	if return_pil:
		return image
	else:
		buffered = BytesIO()
		image.save(buffered, format=image_format)
		img_b64_str = base64.b64encode(buffered.getvalue()).decode()
		return img_b64_str
```

```llama.cpp/examples/llava/clip.cpp
bool clip_image_preprocess(struct clip_ctx * ctx, const clip_image_u8 * img, clip_image_f32 * res, const bool pad2square) {
	if (!ctx->has_vision_encoder) {
		printf("This gguf file seems to have no vision encoder\n");
		return false;
	}

	// the logic below is to pad the shorter side to the longer side with a background color: rgb(122, 116, 104)
	// see https://github.com/haotian-liu/LLaVA/blob/e854a2bf85118c504f6f16bf5c3c7c92f8fa8c6b/llava/conversation.py#L113-L156

	clip_image_u8 * temp = clip_image_u8_init(); // we will keep the input image data here temporarily
	if (pad2square && img->nx != img->ny) {
		int longer_side = std::max(img->nx, img->ny);
		temp->nx = longer_side;
		temp->ny = longer_side;
		temp->buf.resize(3 * longer_side * longer_side);
		const uint8_t bc[3] = {122, 116, 104}; // background color in RGB from LLaVA

		// fill with background color
		for (size_t i = 0; i < temp->buf.size(); i++) {
			temp->buf[i] = bc[i % 3];
		}

		// copy from the input image
		for (int y = 0; y < img->ny; y++) {
			for (int x = 0; x < img->nx; x++) {
				const int i = 3 * (y * img->nx + x);
				const int j = 3 * (y * temp->nx + x);
				temp->buf[j]   = img->buf[i];
				temp->buf[j+1] = img->buf[i+1];
				temp->buf[j+2] = img->buf[i+2];
			}
		}
	} else {
		temp->nx = img->nx;
		temp->ny = img->ny;
		temp->buf.resize(img->buf.size());
		memcpy(temp->buf.data(), img->buf.data(), temp->buf.size());
	}

	const int nx = temp->nx;
	const int ny = temp->ny;

	const int nx2 = ctx->vision_model.hparams.image_size;
	const int ny2 = ctx->vision_model.hparams.image_size;

	res->nx = nx2;
	res->ny = ny2;
	res->buf.resize(3 * nx2 * ny2);

	const float scale = std::max(nx, ny) / (float)ctx->vision_model.hparams.image_size;

	const int nx3 = int(nx / scale + 0.5f);
	const int ny3 = int(ny / scale + 0.5f);

	const auto & m3 = ctx->image_mean; // {0.48145466f, 0.4578275f, 0.40821073f};
	const auto & s3 = ctx->image_std;  // {0.26862954f, 0.26130258f, 0.27577711f};

	for (int y = 0; y < ny3; y++) {
		for (int x = 0; x < nx3; x++) {
			for (int c = 0; c < 3; c++) {
				// linear interpolation
				const float sx = (x + 0.5f) * scale - 0.5f;
				const float sy = (y + 0.5f) * scale - 0.5f;

				const int x0 = std::max(0, (int)std::floor(sx));
				const int y0 = std::max(0, (int)std::floor(sy));

				const int x1 = std::min(x0 + 1, nx - 1);
				const int y1 = std::min(y0 + 1, ny - 1);

				const float dx = sx - x0;
				const float dy = sy - y0;

				const int j00 = 3 * (y0 * nx + x0) + c;
				const int j01 = 3 * (y0 * nx + x1) + c;
				const int j10 = 3 * (y1 * nx + x0) + c;
				const int j11 = 3 * (y1 * nx + x1) + c;

				const float v00 = temp->buf[j00];
				const float v01 = temp->buf[j01];
				const float v10 = temp->buf[j10];
				const float v11 = temp->buf[j11];

				const float v0 = v00 * (1.0f - dx) + v01 * dx;
				const float v1 = v10 * (1.0f - dx) + v11 * dx;

				const float v = v0 * (1.0f - dy) + v1 * dy;

				const uint8_t v2 = std::min(std::max(std::round(v), 0.0f), 255.0f);

				const int i = 3 * (y * nx3 + x) + c;

				res->buf[i] = ((float(v2) / 255.0f) - m3[c]) / s3[c];
			}
		}
	}
	clip_image_u8_free(temp);

	return true;
}
```

cmp-nct commented:
> If it is just the pre-processing missing, we can merge this and make a separate issue with a specific goal to implement that pre-processing. Might be a good idea to first confirm that using a correctly pre-processed image (from the reference implementation) yields good results using this code

PR:
I'm investing more time into it already and I expect a more complete PR soon, I also found issues in the 1.5 implementation yesterday which will get covered alongside and should increase llava-1.5.
I expect to have something within this week.

Testing:
The preprocessing involves more than image modifications, you can not preprocess an image outside and feed it in (you'd need to feed it the embeddings for such a test). 
With llava-1.6 we actually create up to 5 images out of one source image, process those spatially separated in ViT and then feed those projected embeddings into the LLM.

cmp-nct commented:
I'm almost done but got stuck for today on the 5 dimensional permutations that arrange the final embeddings. 

I tried to created an own slim tensor manipulation class, it's probably buggy.
ggml sadly doesn't support 5 dimensions, maybe someone knows a slim approach for operations like these ?
```
        Tensor tensor({combined_embeddings.size()});
        tensor.set_data(combined_embeddings);
        std::vector<int> new_shape = {grid_shape.second, grid_shape.first, num_patches_per_side, num_patches_per_side, clip_n_mmproj_embd(ctx_clip)};
        tensor.view(new_shape);
        tensor.permute({0, 2, 1, 3, 4});
        tensor.contiguous();
        tensor.flatten(0, 3);
        tensor.concatenate(base_tensor, 0);
        std::vector<size_t> new_shape_1d = { tensor.shape[0] * tensor.shape[1] * tensor.shape[2] * tensor.shape[3] * tensor.shape[4]};
        tensor.view(new_shape_1d);
```

ggerganov commented:
Since dims 3 and 4 or not permuted, you can reshape to 4d tensor, apply permute + cont and then reshape back to 5d tensor

cmp-nct commented:
I'm cleaning up the code and hope a first PR update by tomorrow.
Solved: image grid embeddings with newline embedding, bicubic resize, normalization bugfix (1.5 too)
I'm currently not permuting the grid embeddings into patches, while I was able to solve the 5D limitation for the (non unpad) variant thanks to gg's hint, I am still not getting correct results.
The permutation in python results in different code than in ggml, super frustrating that cost me a full day by now.

Even without the grid-embeddings permutation I am getting results better than any other llava variant I tested.


Below are 34B @3 bit (llm) and 6bit (ViT) quantized results, I'm not satisfied until the permutation works but it's anyway quite good. It's about GPT4V/Cog-VLM level but with the permutation done it will exceed it.

![image](https://github.com/ggerganov/llama.cpp/assets/78893154/af543e68-7c7a-48e5-88ff-d5127de563e9)

`The image shows a person standing on the back of a vehicle, which is an unusual and dangerous situation. This is not a typical way to ride in a car or truck, as it poses significant risks for injury due to lack of safety restraints and exposure to potential accidents. The individual appears to be ironing clothes while standing on the vehicle, which adds to the absurdity of the scene, as this activity would normally take place indoors and not while being transported down a street. Additionally, there are other vehicles in motion nearby, which further highlights the hazardous nature of this situation. It's important to note that such behavior is illegal and can lead to severe consequences for both the person involved and others on the road.`

![image](https://github.com/ggerganov/llama.cpp/assets/78893154/01591769-fe41-465f-aa7a-9fc854606a41)

`This image shows three cats sharing a meal from two bowls placed on the ground. The cat in the foreground is black and appears to be eating from one of the bowls, while another cat with a mix of colors (possibly calico) is also eating from the same bowl. A third cat, which seems to have a similar color pattern as the second but is partially obscured by the other two, is peeking out from behind them. The setting appears to be an outdoor area with a concrete surface and what looks like a wall or structure in the background. It's a candid moment that captures the cats' behavior during feeding time. The image has a shallow depth of field, which blurs the background and keeps the focus on the cats and their immediate surroundings.`

![image](https://github.com/ggerganov/llama.cpp/assets/78893154/33f8b2fe-7ab1-4cf0-900a-ad6da6e639c7)

```
{
  "text": [
    "CALIFORNIA USA DRIVER LICENSE",
    "DL 12345678",
    "EXP 08/31/2014",
    "END NONE",
    "CLASS C",
    "LN CARDHOLDER FN IMA",
    "2570 24TH STREET ANYTOWN, CA 95618",
    "DOB 08/31/1977",
    "VETERAN",
    "SEX F",
    "HAIR BRN",
    "EYES BRN",
    "HT 5'6",
    "WT 125 LB",
    "ISS DD 00000000/ANYNN/ANFDYY",
    "08/31/2009"
  ]
}
```

cmp-nct committed:
Clip: Bugfix for normalization (it did not loat the 3 std and mean values)
Clip: bicubic resize function
Clip: added save-to-bmp/pil for debugging and conversion from/to 32/8 images
Clip: added normalization with FP16 precision simulation (image tensors match HF implementation, can be switched off, only used for llava-1.6)
Clip: added newline tensor, mergetype kv, image-grid kv, new resize-pad function with resolution from gridpoints
Clip: clip_image_preprocess now returns a float * vector instead of float, this way llava 1.5 and 1.6 is supported
llava: added ggml cpu graph for embedding patching, added spatial_unpad preliminary support, added a lot of comments that need to be cleaned when all is final
convert-image-encoder: fixed image-grid flattening

cmp-nct commented:
I just pushed, I was not able to finalize it completely and will be mostly busy for the weekend.
**This  PR now includes:**

- Clip: Bugfix for normalization (it did not loat the 3 std and mean values, so all llava-1.5 image tensors were with incorrect normalization)
- Clip: bicubic resize function
- Clip: added save-to-bmp/pil for debugging and conversion from/to 32/8 images
- Clip: added normalization with FP16 precision simulation (image tensors match HF implementation, can be switched off, only used for llava-1.6)
- Clip: added newline tensor, mergetype kv, image-grid kv, new resize-pad function with resolution from gridpoints
- Clip: clip_image_preprocess now returns a float * vector instead of float, this way llava 1.5 and 1.6 is supported
- llava: added ggml cpu graph for embedding patching, added spatial_unpad preliminary support, added a lot of comments that need to be cleaned when all is final
- convert-image-encoder: fixed image-grid flattening

I hope everything is compiling, it's 9AM so I could not test the PR.
It's a large jump in generation quality (as seen in the image demos above) but there still is one problem, I think it's with the tensor strides and would super appreciate if @ggerganov or @slaren maybe could take a look as you've much more tensor knowhow than me: as soon as the permutation of the grid-image embeddings (see at bottom) is applied it fails to inference properly. Without the permutation we get the above (quite good) results.

I have simplified the original python implementation due to the lack of 5 dimensional tensors in GGML, I tested it on the python and it did not result in a noticeable output quality drop.

I am uploading the quantized projectors again on HF, they need an update. the gguf of the LLMs are still compatible.
https://huggingface.co/cmp-nct/llava-1.6-gguf/tree/main
**To create a clip+projector yourself, you need to add this into the vision config part:**
```
  "image_aspect_ratio": "anyres",
  "image_crop_resolution": 224,
  "image_grid_pinpoints": [
    [
      336,
      672
    ],
    [
      672,
      336
    ],
    [
      672,
      672
    ],
    [
      1008,
      336
    ],
    [
      336,
      1008
    ]
  ],
  "mm_patch_merge_type": "spatial_unpad",
  "mm_projector_lr": null,
  "mm_projector_type": "mlp2x_gelu",
```

**Running:**
The llava-1.6 34B model creates very good results, in my tests GPT-4V level (with this implementation despite the hotfix)
You need a lot of context for the additional embeddings, except for that it's normal llava:
`llava-cli.exe -m Q:\models\llava\llava-v1.6-vicuna-7b\ggml-model-f16-q_5_k.gguf --mmproj Q:\models\llava\llava-v1.6-vicuna-7b\mmproj-model-f16.gguf --image C:\temp\LICENSE_DEMO.jpg -p "Provide a full description in JSON format" -ngl 50 --temp 0 -n 500 -ngl 50  -c 4000`
Mistral:
`-e -p "<image>\nUSER:\nProvide a full description in JSON format\nASSISTANT:\n"`
Yi-34B:
`-p "<|im_start|>system\nAnswer the questions.\n\n<image>\n<|im_start|>user\nWhat can be said about this image?\n<|im_start|>assistant\n" -e`




**General problems:**
1) The embedded ViT is different than the vanilla ViT they claim to use, I get different embeddings with each.
2) I was not able to replicate the HF-CLIP embeddings (using a fully black image), this is a general issue with the ggml clip implementation, input tensors are exactly the same, output differs (might be not a problem though).
3) Current CLIP can not be run in batched mode when using a mmprojector, I think we could just extract the llm-projector into a secondary sequential graph for llava-1.6. Then batching could be used for the image-grid. - A 5-image grid (max) inferences in 150ms on my GPU, so it's quite fast without batching.

**Implementation problems:**
1) The permutation fails, I have added the below line of code into the llava.cpp handle_patches() graph which just reverses the permutation again. So the grid-embeddings are added without the 24x24 patching. I think it's a strides error.
2) GGML 5-dim tensor support would make it possible to run the unpad() routine which makes it more efficient

Here is the current Hotfix in llava.cpp to reverse the permutation again:
```
permuted_cont = ggml_cont(model.ctx, ggml_permute(model.ctx, permuted_cont, 0, 2, 1, 3)); // permute back to before - todo: fix bug
```



**Details on the modification of the original implementation:**
```
    // Python reference for full unpad is basically:
        // base_image_feature = image_feature[0]
        // image_feature = image_feature[1:]
        // image_feature = image_feature.permute(4, 0, 2, 1, 3).contiguous()
        // image_feature = image_feature.flatten(1, 2).flatten(2, 3)
        // image_feature = unpad_image(image_feature, image_sizes[image_idx])
        // image_feature = torch.cat((
        //     image_feature,
        //     self.model.image_newline[:, None, None].expand(*image_feature.shape[:-1], 1)
        // ), dim=-1)
        // image_feature = image_feature.flatten(1, 2).transpose(0, 1)
        // image_feature = torch.cat((base_image_feature, image_feature), dim=0)
               
        Pytorch reference simplified, modified for ggml compatibility - confirmed identical output in python (for a 2x2 grid image (676x676 scaling))
        # original without unpadding:
        # image_feature = image_feature.view(2, 2, 24, 24, 4096)
        # image_feature = image_feature.permute(0, 2, 1, 3, 4).contiguous()
        # image_feature = image_feature.view(2, 24, 2, 24, 4096)
        # image_feature = image_feature.flatten(0, 3)

        # Final version in 4D without unpadding:
        Reshape to 4D tensor by merging the last two dimensions
        image_feature = image_feature.view(2, 2, 24, 24*4096)
        image_feature = image_feature.permute(0, 2, 1, 3).contiguous()
        image_feature = image_feature.view(-1, 4096)
```

haotian-liu commented:
WOW thank you so much for implementing LLaVA-1.6 in llama.cpp!!!

One quick note about the prompt for our 34b model: Ideally the correct format should be

```
<|im_start|>system\nAnswer the questions.<|im_end|><|im_start|>user\n<image>\nWhat can be said about this image?<|im_end|><|im_start|>assistant\n
```

And please let me know if there is anything I could do to help (but not about tensor in cpp😭).

slaren commented:
I don't know enough about this architecture to comment, but the `ggml_permute` operation does not operate in the same way as with pytorch, the indices represent the dimensions in the output tensor rather than in the source tensor, which is the opposite as in pytorch. I am not sure why it is that way.

cmp-nct commented:
> WOW thank you so much for implementing LLaVA-1.6 in llama.cpp!!!
> 
> One quick note about the prompt for our 34b model: Ideally the correct format should be
> 
> ```
> <|im_start|>system\nAnswer the questions.<|im_end|><|im_start|>user\n<image>\nWhat can be said about this image?<|im_end|><|im_start|>assistant\n
> ```
> 
> And please let me know if there is anything I could do to help (but not about tensor in cpp😭).

Thanks for your kind words and amazing work, I've been a fan of your elegant approach to image llm's from the start.
I have a quick few questions if you have time:
1) Why do you do the permutation of the grid output embeddings into the 24 segments and not use them as they are ?
2) Is it intentional that there is no newline token between base and grid embeddings ?
3) Is the embedded ViT a fine-tuned variant or the vanilla openAI version ?

haotian-liu commented:
Thank you for your kind words as well!!

The reason is to deal with images with different aspect ratios. I draw the figure below to better illustrate.

<img width="1466" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/6631389/af9af80d-afb6-48dc-8294-327a14290908">

> Why do you do the permutation of the grid output embeddings into the 24 segments and not use them as they are ?
The clip feature embedding for each grid is 576 tokens (24x24), and to merge them into a large feature, and also allow the model to understand what the shape is, we did this spatial unpad and permute thing.

> Is it intentional that there is no newline token between base and grid embeddings ?
The new line token is mainly used to indicate the shape for the model, and the base feature will always be the same shape, so there is no need.

> Is the embedded ViT a fine-tuned variant or the vanilla openAI version ?
We finetuned (unfreeze) ViT backbone during the training, which helps with OCR the most.

Please let me know if there are other confusions. Thank you!

cmp-nct commented:
> I don't know enough about this architecture to comment, but the `ggml_permute` operation does not operate in the same way as with pytorch, the indices represent the dimensions in the output tensor rather than in the source tensor, which is the opposite as in pytorch. I am not sure why it is that way.

Thanks, though I think it's not the problem. I think I'm stuck with a misunderstanding.
The permutation is probably still compatible given it is 0, 2, 1, 3
Would you be on discord for a short chat during the next week maybe ?


> I don't know enough about this architecture to comment, but the `ggml_permute` operation does not operate in the same way as with pytorch, the indices represent the dimensions in the output tensor rather than in the source tensor, which is the opposite as in pytorch. I am not sure why it is that way.

I didn't know that, thanks! ggml needs a fundamental documentation so much.
Though the permutation is "0, 2, 1, 3", so it would be compatible.
Are you possibly on Discord for a short chat when we both have time during the next week ?

haotian-liu commented:
@cmp-nct 😅 your response above is a bit repetitive (not sure if it is due to a bug) and is a bit confusing. Are you responding to me or @slaren?

cmp-nct commented:
> @cmp-nct 😅 your response above is a bit repetitive (not sure if it is due to a bug) and is a bit confusing. Are you responding to me or @slaren?

I'm sorry, it was to slaren :)

Your explanation was helpful, thanks! Except for the ViT part (if the embedded ViT is the original vanilla openAI ViT or if it was finetuned like you did with ShareGPT)

haotian-liu commented:
> Except for the ViT part (if the embedded ViT is the original vanilla openAI ViT or if it was finetuned like you did with ShareGPT)

The architecture is the OpenAI ViT and there is ZERO modifications to the arch. We do not specifically finetune it for any tasks, but we unfreeze it and it's optimized with the gradients coming from our masked language modeling loss.

ggerganov commented:
> the ggml_permute operation does not operate in the same way as with pytorch, the indices represent the dimensions in the output tensor rather than in the source tensor, which is the opposite as in pytorch. I am not sure why it is that way.

No particular reason - it was implemented in a way that made sense to me at the time. I wasn't aware there is a permute function in pytorch

Commit:
whitespace corrections

Commit:
ws

Commit:
Tensors are now properly permuted.
Before the embeddings were inserted 1:1, now they are split into the 24x24 patches as in reference.

Commit:
ws

Commit:
added verbose_prompt support into cli
added stopwords for llava-1.6 into cli

cmp-nct commented:
Been too long since I did ggml tensors, so I had forgotten that the dimensional representation is inverse to pytorch.
Cost me a couple gray hair, that really should be mentioned somewhere ;)
**I've marked the PR ready for review.**

Below are the two llava reference images, I ran them on LLM-Q5K and CLIP-Q6K using this PR
![image](https://github.com/ggerganov/llama.cpp/assets/78893154/2fe07826-d5b0-4b09-9206-0f00f810dc52)
Reference: `The image shows a person ironing clothes on a portable ironing board while standing on the back of a moving vehicle, which is unusual and unsafe. This is not a typical or safe way to iron clothes, as it exposes the person to the risk of falling off the vehicle, being hit by other vehicles, or being injured by the iron itself. Additionally, the person is in the middle of a busy street, which adds to the danger of the situation.` 

This PR: `The image shows a person sitting on the back of a vehicle, which is an unusual and potentially dangerous situation. It's not common to see someone in such a precarious position while a vehicle is moving or parked, as it poses significant risks for both the individual and others around them. The person appears to be using an ironing board and ironing clothes, which adds to the absurdity of the scene because this activity would typically take place indoors and not on the back of a moving vehicle. Additionally, there are other vehicles in the background that suggest this is happening in a busy urban area, further emphasizing the unusual and hazardous nature of the situation depicted in the image.` 


![image](https://github.com/ggerganov/llama.cpp/assets/78893154/42c8156a-4988-40df-9ac9-bdbed748c516)


```
When visiting a location like the one shown in the image, which appears to be a serene lake with a dock and surrounded by forested mountains, there are several things to be cautious about:

1. Water Safety: If you plan to swim, make sure you're aware of the water conditions, such as currents, depth, and temperature. Wear a life jacket if you're not a strong swimmer.
2. Weather Conditions: Be prepared for changes in weather. Mountain weather can be unpredictable, so check the forecast and dress appropriately.
3. Wildlife: Be aware of the local wildlife. Do not feed animals, and keep a safe distance.
4. Terrain: The dock and surrounding areas may be slippery, especially if there's algae growth or if it's wet. Wear appropriate footwear with good traction.
5. Personal Safety: If you're alone, let someone know where you're going and when you expect to return. It's also a good idea to carry a charged mobile phone in case of emergencies.
6. Leave No Trace: Respect the environment by not littering and by staying on designated paths or areas.
7. Regulations: Follow any posted signs or regulations, which may include fishing licenses, campfire rules, or restrictions on certain activities.
8. Equipment: Bring the right equipment for your activities, such as binoculars for birdwatching, a camera for photography, or a fishing rod if you're fishing.
9. Insects: Depending on the season, there may be insects like mosquitoes or ticks. Use insect repellent and wear long sleeves and pants to protect your skin.
10. Emergency Preparedness: Have a basic first aid kit and know the location of the nearest medical facility.

Enjoying nature responsibly and being prepared for various situations will help ensure a safe and enjoyable visit.
```

This PR:
```
When visiting a location like the one shown in the image, which appears to be a serene lake with a dock and surrounded by mountains and forest, there are several things you should be cautious about:

1. Water Safety: If you plan on swimming or boating, make sure you're aware of water safety rules. Wear a life jacket if you're not a strong swimmer or if you'll be in a boat. Be mindful of the water temperature and currents.

2. Weather Conditions: Mountain weather can change rapidly. Check the forecast before you go and be prepared for sudden changes, such as rain, snow, or fog.

3. Wildlife: Depending on the region, there may be wildlife to watch out for. Keep a safe distance from any animals you encounter and do not feed them.

4. Terrain: Be cautious of slippery surfaces on the dock, especially if it's wet. Watch your step when walking around the lake or in the forest as there might be uneven terrain or hidden obstacles.

5. Personal Safety: If you're alone, let someone know where you are going and when you expect to return. It's always safer to have a companion when exploring natural areas.

6. Leave No Trace: Respect the environment by not littering and by staying on designated paths or trails if there are any.

7. Equipment: Bring appropriate gear for your activities, such as sturdy footwear, sunscreen, insect repellent, and a first aid kit.

8. Communication: Have a way to communicate in case of an emergency. Cell phone reception may be unreliable in remote areas, so consider alternative methods like a satellite phone or a personal locator beacon.

9. Local Regulations: Familiarize yourself with any local regulations regarding fishing, camping, and other activities you might want to do.

10. Altitude Sickness: If the lake is at a high altitude, be aware of the symptoms of altitude sickness and take precautions such as acclimatizing properly and staying hydrated.

Remember that these are general guidelines, and specific advice may vary depending on the exact location and your personal plans for the visit.
```

chigkim commented:
Do people who quantized the model need to quantize again with the latest commit from PR?

ggerganov commented on examples/llava/clip.h:
```diff
@@ -3,6 +3,8 @@
 
 #include <stddef.h>
 #include <stdint.h>
+#include <string>
+#include <vector>
```
We should try to keep the CLIP API C-style

ggerganov commented on examples/llava/clip.h:
```diff
@@ -44,6 +56,24 @@ CLIP_API size_t clip_embd_nbytes(const struct clip_ctx * ctx);
 CLIP_API int clip_n_patches    (const struct clip_ctx * ctx);
 CLIP_API int clip_n_mmproj_embd(const struct clip_ctx * ctx);
 
+// RGB uint8 image
+CLIP_API struct clip_image_u8 {
+    int nx;
+    int ny;
+
+    std::vector<uint8_t> buf;
+};
+
+// RGB float32 image (NHWC)
+// Memory layout: RGBRGBRGB...
+ CLIP_API struct clip_image_f32 {
+    int nx;
+    int ny;
+
+    std::vector<float> buf;
+};
+
```
These structs should not be exposed through the API

ggerganov commented on examples/llava/clip.cpp:
```diff
@@ -1134,13 +1170,423 @@ bool clip_image_load_from_bytes(const unsigned char * bytes, size_t bytes_length
+    }
+
+    file.close();
+}
```
I guess these functions are used just for debugging? Better to remove them

ggerganov commented on examples/llava/clip.cpp:
```diff
@@ -1134,13 +1170,423 @@ bool clip_image_load_from_bytes(const unsigned char * bytes, size_t bytes_length
+
+// for replication purposes `.to(model.device, dtype=torch.float16)`
+// converts a float to half precision and back to float
+float simulateFloat16Precision(float value) {
```
It's probably better to utilize the existing `ggml_fp16_to_fp32_row` and `ggml_fp32_to_fp16_row` for this task

ggerganov commented on examples/llava/clip.cpp:
```diff
@@ -1134,13 +1170,423 @@ bool clip_image_load_from_bytes(const unsigned char * bytes, size_t bytes_length
 // TODO: implement bicubic interpolation instead of linear.
-bool clip_image_preprocess(struct clip_ctx * ctx, const clip_image_u8 * img, clip_image_f32 * res, const bool pad2square) {
+// returns the normalized float tensor for llava-1.5, for spatial_unpad with anyres processing for llava-1.6 it returns the normalized image patche tensors as a vector
+bool clip_image_preprocess(struct clip_ctx * ctx, const clip_image_u8 * img, std::vector<clip_image_f32*>& res_tensor, bool pad2square) {
```
`res_tensor` -> `res_imgs`

`pad2square` -> `pad_to_square`

Commit:
moved llava functions to llava.cpp, made clip.h C compatible API, replaced vector style functions with pointers, added a debug define to remove functions from compilation while not needed

cmp-nct commented:
> Do people who quantized the model need to quantize again with the latest commit from PR?
No, all you need is the right projector

`encode_image_with_clip: image embedding created: 576 tokens` < this is llava-1.5
`encode_image_with_clip: image embedding created: 2880 tokens` < this is llava-1.6 (anything above 576)

Commit:
ws

jpohhhh commented:
Wow - I was using 1.6 GGUFs with the old CLiP and was already stunned. Really looking forward to this, thank you.

Double-checking and confirming: an image will now consume 2880 tokens of the LLM model's context?

ggerganov commented:
I'm trying to convert LLaVA 1.6:

- downloaded https://huggingface.co/liuhaotian/llava-v1.6-mistral-7b/tree/main
- ran `python3 examples/llava/llava-surgery-v2.py -m ~/Data/huggingface/llava-v1.6-mistral-7b` successfully
- output says:

```
Also, use /Users/ggerganov/Data/huggingface/llava-v1.6-mistral-7b/llava.projector to prepare a llava-encoder.gguf file.
```

How to do that?

ggerganov committed:
convert : skip unknown tensors (need for LLaVA)

ggerganov committed:
llava : update readme

ggerganov commented:
@cmp-nct I would like to push some fixes - need permission to the branch:

```
$ git push cmp-nct HEAD:llava-1.6_1 
remote: Permission to cmp-nct-org/llama.cpp.git denied to ggerganov.
fatal: unable to access 'https://github.com/cmp-nct-org/llama.cpp/': The requested URL returned error: 403
```

evq commented on examples/llava/clip.cpp:
```diff
@@ -1134,24 +1190,378 @@ bool clip_image_load_from_bytes(const unsigned char * bytes, size_t bytes_length
+    int original_height = original_size.second;
+    std::pair<int, int> best_fit;
+    int max_effective_resolution = 0;
+    int min_wasted_resolution = std::numeric_limits<int>::max();
```
I had to add an `#include <limits>` otherwise I hit this:
```
examples/llava/clip.cpp:1475:38: error: 'numeric_limits' is not a member of 'std'
 1475 |     int min_wasted_resolution = std::numeric_limits<int>::max();
      |                                      ^~~~~~~~~~~~~~
```

ggerganov committed:
llava : fix compile warnings

ggerganov committed:
Merge remote-tracking branch 'origin/master' into HEAD

ggerganov committed:
llava : style

ggerganov committed:
convert : add --skip-unknown CLI arg

ggerganov commented:
Yup, the state of the clip/llava implementation is not great - hopefully nobody uses this in production

I've added `--skip-unknown` to `convert.py`

ggerganov committed:
server : remove clip structs

ggerganov reviewed (approved):
Should be ok to merge - let's give it some time until tomorrow and if no issues are reported we can merge it

cmp-nct commented:
Thanks, looking forward to see a closure on that one. It was a big pain to get working :-)

cmp-nct committed:
bugfix for non llava-1.6

It should now work with llava-1.5 as well

ggerganov committed:
clip : minor code rearrange

ggerganov committed:
llava : update readme a bit

ggerganov commented:
It would be very useful to add detailed instructions for LLaVA v1.6 in the README. I tried writing them, but I realized I don't know which is the correct CLIP model to use - I think I'm using the small one with 576 tokens embedding. Not sure

Anyway, if anyone figures out all the steps, please add open a PR

cmp-nct commented:
I've updated the readme with detailed instructions and hints on llava-1.6

In general: I have used llava-surgery-v2 only a little bit outside llava-1.6, it is meant as a full replacement to llava-surgery.py.
It should now work with all models, safetensor as well as pytorch and with projectors and visual tensors in any locations but I've not tested it in great detail on non llava-1.6 yet.

Maybe after a while the original llava-surgery can be removed, once we know the new one works for everything

Elbios commented:
Hi all, I tried running this on latest master:

./bin/server -m ../models/mistral-7b-q_5_k.gguf --mmproj ../models/mmproj-mistral7b-f16-q6_k.gguf  -ngl 50  -c 6000 --host 0.0.0.0 --port 8007 --no-mmap

and got:

llama_new_context_with_model: graph splits (measure): 3
Available slots:
 -> Slot 0 - max context: 6000
{"timestamp":1707926446,"level":"INFO","function":"main","line":2623,"message":"model loaded"}
all slots are idle and system prompt is empty, clear the KV cache
slot 0 - loaded image
slot 0 is processing [task id: 0]
slot 0 : kv cache rm - [0, end)
slot 0 - encoding image [id: 1]
munmap_chunk(): invalid pointer
Aborted

and traced it down to a memory management error. Fixed it here:

https://github.com/ggerganov/llama.cpp/pull/5491

Great work on this btw, llava 1.6 is fantastic!

jpohhhh commented:
Similarly, possible fix for bad access crash on wide images: https://github.com/ggerganov/llama.cpp/pull/5493

And got the same impression in testing: definitely a huge step up :)

chigkim commented:
@ggerganov @cmp-nct This is amazing! Thanks!
Does this work on main cli as well as server, or someone has to modify them?
Main cli has --mmproj and --image options, and server has api to send  image as well as web UI to upload an image in the chat mode.
What's the difference between using it with main and llava-cli other than llava-cli gives much fewer options?

cmp-nct commented:
> @ggerganov @cmp-nct This is amazing! Thanks! Does this work on main cli as well as server, or someone has to modify them? Main cli has --mmproj and --image options, and server has api to send image as well as web UI to upload an image in the chat mode. What's the difference between using it with main and llava-cli other than llava-cli gives much fewer options?

It should work on both llava-cli and the server.
I personally only tested it on llava-cli in it's latest iteration
CLIP does not come with many options, so both are equal

chigkim commented:
Just curious...
Why does llava-cli exist separately when main cli has --mmproj and --image flags?
Wouldn't it be better to fold llava-cli into main cli?

cmp-nct commented:
> Just curious... Why does llava-cli exist separately when main cli has --mmproj and --image flags? Wouldn't it be better to fold llava-cli into main cli?

I've not actually tested but last time I looked main did not support that.
The command line arguments are in "common" and they are not modularized, so every command line argument is shared with every example. You also have like 40 command line options when using llava-cli -h, barely any of those will have an effect.

chigkim commented:
Thanks so much for the explanation . You're totally right. I just tried main, and it doesn't work.

I got llava-cli with llava 1.6 13B to work on free Colab t4 , and hopefully someone can fix server to work properly with llava 1.6. That would be amazing!

https://colab.research.google.com/gist/chigkim/c44dcc37af26f1cb3af03a2209d7c50a/llava16.ipynb#scrollTo=_DqzhwSz_eMB

More importantly, I believe Ollama also uses llama.cpp server.

jxy commented:
I had two issues with the `convert-image-encoder-to-gguf.py`.
1. The correct argument seems to be `--clip-model-is-vision` instead of the one with underscore `_` in README.
2. I used `https://huggingface.co/cmp-nct/llava-1.6-gguf/resolve/main/config.json` but I needed to change the python file
```diff
diff --git a/examples/llava/convert-image-encoder-to-gguf.py b/examples/llava/convert-image-encoder-to-gguf.py
index c69f89ac..94754a47 100644
--- a/examples/llava/convert-image-encoder-to-gguf.py
+++ b/examples/llava/convert-image-encoder-to-gguf.py
@@ -117,7 +117,7 @@ else:
 with open(dir_model + "/config.json", "r", encoding="utf-8") as f:
     config = json.load(f)
     if args.clip_model_is_vision:
-        v_hparams = config
+        v_hparams = config["vision_config"]
         t_hparams = None
     else:
         v_hparams = config["vision_config"]
```

cmp-nct commented:
> I had two issues with the `convert-image-encoder-to-gguf.py`.
> 
> 1. The correct argument seems to be `--clip-model-is-vision` instead of the one with underscore `_` in README.
> 2. I used `https://huggingface.co/cmp-nct/llava-1.6-gguf/resolve/main/config.json` but I needed to change the python file
> 
> ```diff
> diff --git a/examples/llava/convert-image-encoder-to-gguf.py b/examples/llava/convert-image-encoder-to-gguf.py
> index c69f89ac..94754a47 100644
> --- a/examples/llava/convert-image-encoder-to-gguf.py
> +++ b/examples/llava/convert-image-encoder-to-gguf.py
> @@ -117,7 +117,7 @@ else:
>  with open(dir_model + "/config.json", "r", encoding="utf-8") as f:
>      config = json.load(f)
>      if args.clip_model_is_vision:
> -        v_hparams = config
> +        v_hparams = config["vision_config"]
>          t_hparams = None
>      else:
>          v_hparams = config["vision_config"]
> ```

The readme was updated but I think just merged today.
The config I had in the HF was for a full model but not for the extracted ViT. The current readme points to the right file, so your modification won't be necessary.

The entire llava-1.6 change including the last minute refactors were a bit much for me to push at once, I've had dozens of variants local, that's how that issue sneaked in.

At this point llava-cli appears to work flawless, server does not. Server needs an update to use llava.cpp preprocessing functions, it also needs an update to allow flexible system prompt and finetune syntax. It's minor high level work anyone should be able to do.
If no one does it I'll try find time for it next week.

cjpais commented:
> At this point llava-cli appears to work flawless, server does not. Server needs an update to use llava.cpp preprocessing functions, it also needs an update to allow flexible system prompt and finetune syntax. It's minor high level work anyone should be able to do. If no one does it I'll try find time for it next week.

Preprocessing step should be implemented in #5553, thanks for the insight into the problem on the other threads. 

> it also needs an update to allow flexible system prompt and finetune syntax

Could you explain more on this, happy to try to include it as well

arkohut commented:
![screenshot2024-02-20 18 32 12](https://github.com/ggerganov/llama.cpp/assets/39525455/7d76a44b-592b-4e91-9daf-175f3cbbdac9)

I have a question about the context using llava 1.6. Does the screenshot means that the total context of it is `2048` token? It seems that the output even not finished before using out the context.

Here is the command:

```sh
./llava-cli -m ~/ggml-yi-34b-f16-q_5_k.gguf --mmproj ~/mmproj-llava-34b-f16-q6_k.gguf --image ~/00263.jpg -e -p '<|im_start|>system\nAnswer the questions.<|im_end|><|im_start|>user\n<image>\nProvide a full description.<|im_end|><|im_start|>assistant\n'
```

svenstaro commented:
@cjpais Would it be reasonable to update the description on your [HF repo](https://huggingface.co/cjpais/llava-v1.6-34B-gguf)? I think these are probably the most proper quants for LLaVA v1.6 around and fairly complete.

cjpais commented:
@svenstaro, updated
